iree/hal/allocator_caching.c

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/allocator_caching.h"
+
+#include "iree/base/tracing.h"
+#include "iree/hal/detail.h"
+#include "iree/hal/resource.h"
+
+typedef struct iree_hal_caching_allocator_t {
+  iree_hal_resource_t resource;
+  iree_hal_allocator_t* delegate_allocator;
+  iree_string_view_t identifier;
+
+} iree_hal_caching_allocator_t;
+
+static const iree_hal_allocator_vtable_t iree_hal_caching_allocator_vtable;
+
+iree_hal_caching_allocator_t* iree_hal_caching_allocator_cast(
+    iree_hal_allocator_t* base_value) {
+  return (iree_hal_caching_allocator_t*)base_value;
+}
+
+IREE_API_EXPORT iree_status_t iree_hal_allocator_create_caching(
+    iree_string_view_t identifier, iree_hal_allocator_t* delegate_allocator,
+    iree_hal_allocator_t** out_allocator) {
+  IREE_ASSERT_ARGUMENT(out_allocator);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_allocator_cache_buffer = true;
+
+  iree_hal_caching_allocator_t* allocator = NULL;
+  iree_host_size_t total_size =
+      iree_sizeof_struct(*allocator) + identifier.size;
+  iree_status_t status = iree_allocator_malloc(
+      iree_hal_allocator_host_allocator(delegate_allocator), total_size, (void**)&allocator);
+  if (iree_status_is_ok(status)) {
+    
+    iree_hal_resource_initialize(&iree_hal_caching_allocator_vtable,
+                                  &allocator->resource);
+
+    allocator->delegate_allocator = delegate_allocator;
+
+    iree_string_view_append_to_buffer(
+        identifier, &allocator->identifier,
+        (char*)allocator + iree_sizeof_struct(*allocator));
+    
+    *out_allocator = (iree_hal_allocator_t*)allocator;
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+typedef struct iree_hal_caching_allocator_cache {
+  iree_hal_buffer_t *buffer;
+  struct iree_hal_caching_allocator_cache *next;
+} iree_hal_caching_allocator_cache;
+
+bool iree_hal_allocator_cache_buffer;
```

Comment:
I haven't quite tracked through why we need this global, but I suspect once you have your own buffer type, you won't need this. We really need to remove it.